ee/clickhouse/views/person.py

```diff
@@ -1,30 +1,93 @@
 from ee.clickhouse.queries.trends.lifecycle import ClickhouseLifecycle
 from ee.clickhouse.sql.person import GET_PERSON_PROPERTIES_COUNT
 from posthog.api.person import PersonViewSet
-from posthog.api.utils import format_next_url, format_offset_absolute_url
-from posthog.constants import INSIGHT_FUNNELS, FunnelVizType
+from posthog.api.utils import format_offset_absolute_url
+from posthog.constants import FunnelVizType
 from posthog.decorators import cached_function
+from posthog.exceptions import UnsupportedFeature
 from posthog.models import Event, Filter, Person
 
+BASE_FILTER_PERSONS_QUERY = """
+SELECT person.id FROM person
+LEFT OUTER JOIN (
+    SELECT * FROM person_distinct_id WHERE person_distinct_id.team_id = %(team_id)s
+) person_distinct_id
+ON person.id = person_distinct_id.person_id
+WHERE team_id = %(team_id)s"""
+
+
+def filter_persons_ch(team_id: int, request: Request, queryset: QuerySet) -> QuerySet:
+    # Keep functionality in sync with posthog/api/person.py
+    params: Dict[str, Any] = {"team_id": team_id}
+    and_conditions: List[str] = []
+    if request.GET.get("id"):
+        raise UnsupportedFeature("filtering persons by field `id`")
+    if request.GET.get("uuid"):
+        params["uuids"] = request.GET["uuid"].split(",")
+        and_conditions.append("AND has(%(uuids)s, toString(person.id))")
+    if request.GET.get("search"):
+        parts = request.GET["search"].split(" ")
+        contains = []
+        for part_index, part in enumerate(parts):
+            if ":" in part:
+                matcher, key = part.split(":")
+                if matcher == "has":
+                    # Matches for example has:email or has:name
+                    params[f"has_key_{part_index}"] = key
+                    and_conditions.append(f"AND JSONHas(person.properties, %(has_key_{part_index})s)")
+            else:
+                contains.append(part)
+        params["icontains"] = f'%{" ".join(contains)}%'
+        and_conditions.append(
+            f"""
+        AND (
+            person.properties ILIKE %(icontains)s
+            OR person_distinct_id.distinct_id ILIKE %(icontains)s
+        )
+        """
+        )
+    properties: List[Dict[str, Any]] = json.loads(request.GET["properties"]) if request.GET.get("properties") else []
+    if request.GET.get("cohort"):
+        properties.append({"type": "cohort", "key": "id", "value": request.GET["cohort"]})
+    if properties:
+        filter = Filter(data={"properties": properties})
+        filter_query, filter_params = parse_prop_clauses(filter.properties, team_id, is_person_query=True)
+        and_conditions.append(filter_query)
+        params.update(filter_params)
 
-class ClickhousePersonViewSet(PersonViewSet):
+    if and_conditions:
+        final_query = f"""{BASE_FILTER_PERSONS_QUERY} {' '.join(and_conditions)}"""
+        uuids_found_rows = cast(list, sync_execute(final_query, params))
```

Comment:
Q: Why do we need the cast here? Seems wrong.

Reply:
As in some other case we discussed in a different PR, Pylance sees types differently than Mypy (i.e. in this case it's afraid it may be a `None`), but can remove :)